src/app/tests/TestClusterObjectUtils.cpp

```diff
@@ -0,0 +1,782 @@
+        {
+            err = EncodeStruct(_this->mWriter, TLV::AnonymousTag,
+                               MakeTagValuePair(TLV::ContextTag(clusters::TestCluster::NestedStructList::kDFieldId), t.e));
+            NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+        }
+        
+        err = _this->mWriter.Finalize();
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        _this->DumpBuf();
+    }
+
+    //
+    // Decode
+    //
+    {
+        chip::app::clusters::TestCluster::IteratableNestedStructList::Type t;
+
+        _this->SetupReader();
+
+        err = chip::app::ClusterObjectUtils::Decode(_this->mReader, t);
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        auto iter = t.d.begin();
+
+        while (iter != t.d.end()) {
+            NL_TEST_ASSERT(apSuite, iter->HasError());
+            break;
+        }
+    }
+}
+
+
+} // namespace app
+} // namespace chip
+
+namespace {
+
+void InitializeChip(nlTestSuite * apSuite)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::Optional<chip::Transport::PeerAddress> peer(chip::Transport::Type::kUndefined);
+    chip::Transport::FabricTable fabrics;
+
+    err = chip::Platform::MemoryInit();
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    chip::gSystemLayer.Init();
+
+    err = chip::gSessionManager.Init(&chip::gSystemLayer, &chip::gTransportManager, &fabrics, &chip::gMessageCounterManager);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    err = chip::gExchangeManager.Init(&chip::gSessionManager);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    err = chip::app::InteractionModelEngine::GetInstance()->Init(&chip::gExchangeManager, nullptr);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+}
+
+// clang-format off
+const nlTest sTests[] =
+{
+    //NL_TEST_DEF("TestClusterObjectUtils_EncAndDecSimple", chip::app::TestClusterObjectUtils::TestClusterObjectUtils_EncAndDecSimpleStruct),
+    //NL_TEST_DEF("TestClusterObjectUtils_EncAndDecNestedStruct", chip::app::TestClusterObjectUtils::TestClusterObjectUtils_EncAndDecNestedStruct),
+    //NL_TEST_DEF("TestClusterObjectUtils_EncAndDecIterableNestedStructList", chip::app::TestClusterObjectUtils::TestClusterObjectUtils_EncAndDecIterableNestedStructList),
+    //NL_TEST_DEF("TestClusterObjectUtils_EncAndDecIterableDoubleNestedStructList", chip::app::TestClusterObjectUtils::TestClusterObjectUtils_EncAndDecIterableDoubleNestedStructList),
+    //NL_TEST_DEF("TestClusterObjectUtils_OptionalFields", chip::app::TestClusterObjectUtils::TestClusterObjectUtils_OptionalFields),
+    //NL_TEST_DEF("TestClusterObjectUtils_ExtraField", chip::app::TestClusterObjectUtils::TestClusterObjectUtils_ExtraField),
+    //NL_TEST_DEF("TestClusterObjectUtils_InvalidSimpleFieldTypes", chip::app::TestClusterObjectUtils::TestClusterObjectUtils_InvalidSimpleFieldTypes),
+    NL_TEST_DEF("TestClusterObjectUtils_InvalidListType", chip::app::TestClusterObjectUtils::TestClusterObjectUtils_InvalidListType),
```

Comment:
Tests I would like to see:

1. `Span<char>` gets encoded as UTF-8 string, or perhaps does not compile (this last is hard to test; manual test is fine).
2. `Span<const char>` gets encoded as UTF-8 string.
3. `Span<uint8_t>` gets encoded as a list of uint8.
4. `Span<const uint8_t>` gets encoded as an octet string.

Reply:
Can add tests for 1 and 2.

3 and 4 bring up an interesting dilemma. I don't think it's intuitive that the difference between an octstr and list<uint8> is just the const.

I'd rather do this:
```
struct Uint8Span : public Span<const uint8_t> {

};

That would provide a distinction in the types between `chip::ByteSpan`, and `chip::Uint8Span`.
```

Reply:
So the dilemma is that we are already using `Span<const uint8_t>` to represent octet strings, and would like some sort of uniform-ish representation for lists of things, right?  That is, if we introduce a new type to represent lists of `uint8_t` but not new types to represent other lists, it's a bit confusing.